Reduces friction for CUJ "I want to know how to migrate my Manifest V2 extension to Manifest V3" (#2855)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/mv3/intro/mv3-migration/index.md

subhead: 'Getting you headed in the right direction.'
description: 'A high-level guide to how you can migrate your Manifest V2 extensions to Manifest V3.'
date: 2020-11-09
updated: 2022-06-02
---

This guide provides developers with the information they need to begin migrating an extension from
Manifest V2 to Manifest V3. Some extensions will require very little change to make them Manifest V3
compliant, while others will need to be redesigned to some degree. For a quick reference guide see
the [migration checklist][mv3-checklist].

{% Aside %}

Follow [What's new in Chrome Extensions][doc-whats-new] to read about new Manifest V3 features as

> they become available.

Author: I agree! 👍

* [Service workers][mv3-sw] replace background pages.
* [Network request modification][mv3-network] is now handled with the new
  [declarativeNetRequest][mv3-declarative] API.
* [Remotely hosted code][mv3-remote] is no longer allowed; an extension can only execute JavaScript
  that is included within its package.
* [Promise][mv3-promise] support has been added to many methods, though callbacks are still
  supported as an alternative.  (We will eventually support promises on all appropriate methods.)
* A number of other, relatively [minor feature changes][mv3-other] are also introduced in Manifest
  V3.

For a fuller description of these changes, see the [Manifest V3 Overview][mv3-overview].

## Updating the manifest.json file  {: #updating-manifest-dot-json }

To use the features of Manifest V3, you need to update your [manifest file][doc-manifest].
Naturally, you'll change the manifest version to "3", but there are other things you need to change
in the manifest file: the [service worker][section-man-sw], [host permissions][section-host],
[content security policy][section-csp], [action declarations][section-action], and [web-accessible

> resources][section-war].

Author: @dotproto Looks better, thanks for the suggestion :)

```
{% endColumns %}

### Service worker  {: #man-sw }

In Manifest V3, background pages are now *service workers*. Register the service worker under the
`"background"` field. This field uses the `"service_worker"` key, which specifies a single
JavaScript file.

{% Columns %}

```json
// Manifest V2

"background": {
  "scripts": [
    "backgroundContextMenus.js",
    "backgroundOauth.js"
],
  "persistent": false

> }

Author: IMO, adding the ellipsis makes it look a bit bulky... 

<img width="781" alt="Screen Shot 2022-06-10 at 14 34 00" src="https://user-images.githubusercontent.com/37001393/173137615-9f5b9f64-a7f9-4486-9e5f-6af975830e1c.png">

Below is the original code:
<img width="732" alt="Screen Shot 2022-06-10 at 14 36 45" src="https://user-images.githubusercontent.com/37001393/173137935-ee1da975-59e8-4d12-a907-fbab934dc270.png">

> }

Author: @dotproto I added the brackets and ellipsis! I think we are ready to merge

```

```json
// Manifest V3

"background": {
  "service_worker": "background.js",
  "type": "module" //optional
}
```

{% endColumns %}

Even though Manifest V3, does not support multiple background scripts, you can optionally declare
the service worker as an [ES Module][webdev-esm] by specifying `"type": "module"`, which allows you

> to import further code.

Author: Good call. Looks like you already opened an issue for this! Thanks :) #2996

  "http://www.blogger.com/",
],
"optional_permissions": [

>   "*://*/*",

Author: Gotcha! I'll add it back to the manifest V2 code sample

- <code><var>EXTENSION_IDS</var></code>: A list of strings, each containing the ID of a given
  extension.

Previously, the list of web accessible resources applied to all websites and extensions, which
created opportunities for fingerprinting or unintentional resource access. The updated API lets
extensions more tightly control what other sites or extensions can access extension resources. 

See the [web accessible resources][doc-war] documentation for usage information.

## Code execution  {: #code-execution }

Manifest V3 imposes new restrictions that limit an extension's ability to execute unreviewed
JavaScript through a combination of platform changes and policy limitations.

Many extensions are unaffected by this change. However, if your Manifest V2 extension executes
remotely hosted scripts, injects code strings into pages, or evals strings at runtime, you'll need
to update your code execution strategies when migrating to Manifest V3.

> ### Remotely hosted code  {: #remotely-hosted-code }

Author: @dotproto Good call!

_Remotely hosted code_ refers to any code that is not included in an extension's package as a
loadable resource. For example, the following are considered remotely hosted code:

- JavaScript files pulled from a remote server

> - Any library hosted on a [CDN][mdn-cdn].

Author: Great suggestion, and good catch on the period! 😅

```

```html
// Manifest V3

// popup.html
<script src="./react-dom.production.min.js"></script>
<link href="./bootstrap.min.css" rel="stylesheet">
```

{% endColumns %}

To include a library in a service worker, you have two options:
- For standard service workers, use `importScripts()`.
- To use static import statements, set the `"background.type"` to `"module"` in the manifest.

### Executing arbitrary strings  {: #executing-arbitrary-strings }

In Manifest V2, it was possible to execute an arbitrary string of code using
[`tabs.executeScript()`][api-tabs-executescript] and the `code` property on the options object.
Manifest V3 does not allow arbitrary code execution. In order to adapt to this requirement,
extension developers can use the [`scripting.executeScript()`][api-scripting-executescript] API to
either inject a static file or a function.

To use the [Scripting API][api-scripting], you need to include the `"scripting"` permission in your
manifest file. This API does not [trigger a permission warning][doc-perm-warn].

```json
{
  "manifest_version": 3,
  "permissions": ["scripting"],
  ...
}
```

> <web-tabs>

Author: @dotproto I see what you mean, looking further into the tabs component, it's intention is to be used without an intro, only a [single code sample](https://web.dev/handbook/web-dev-components/#tabs).

I'll create an issue to follow-up. Maybe a UX that allows an intro and a border line like in the FB docs:
<img width="843" alt="Screen Shot 2022-06-10 at 16 32 20" src="https://user-images.githubusercontent.com/37001393/173154058-4ea7c819-02be-43fc-8dce-389ee6799b8d.png">

workers as of Chrome 87.

{% endAside %}

## Modifying network requests  {: #modifying-network-requests }

There is a new [declarativeNetRequest][api-declarativenetrequest] for network request modification,
which provides an alternative for much of the [webRequest][api-webrequest] API's functionality.

### When can you use blocking webRequest?  {: #when-use-blocking-webrequest }

The blocking version of the [webRequest][api-webrequest] API still exists in Manifest V3 but its use
is restricted to force-installed extensions only. See Chrome Enterprise policies:
[ExtensionSettings][enterprise-settings], [ExtensionInstallForcelist][enterprise-force-list].

All other extensions must now use [declarativeNetRequest][api-declarativenetrequest] for network
request modification. This moves the actual modification of the network request into the Chrome
browser: the extension no longer can read the actual network request, and in most cases needs no
host permissions.

{% Aside 'caution' %}

Request redirects and header modifications **do** require the user to grant host permissions.

{% endAside %}

### How do you use declarativeNetRequest?  {: #how-use-declarativenetrequest }

> Instead of reading the request and programmatically altering it, your extension specifies a number
> of [rules][api-declarative-rules], which map a set of conditions to corresponding actions. See the
> [declarativeNetRequest][api-declarativenetrequest] reference documentation for a more detailed
> description of rules.

Reviewer: I've apparently entered the late night editing phase where I'm just rewriting everything I read. I'm going to try to stop that now.

```suggestion
Instead of reading the request and programmatically altering it, your extension specifies a number
of [rules][api-declarative-rules]. Each rule contains a set of actions to perform when a given set
of conditions are matched. For example, you could define a rule that removes "cookie" headers when a
request is sent to a specific domain. See the [declarativeNetRequest][api-declarativenetrequest]
reference documentation for a more detailed description of rules
```

This feature allows content blockers and other request-modifying extensions to implement their use
cases without requiring host permissions, and without needing to read the actual requests.

{% Aside %}

In order to aid with the migration process, the declarativeNetRequest API is available for use in
Manifest V2 extensions as of Chrome 84.

> {% endAside %}

Reviewer: This aside feels very out of place here. We definitely should talk about how to effectively use the `"minimum_chrome_version"` field and some key Chrome versions developers might want to target, but this doesn't seem like the right place for it.

### Conditional permissions and declarativeNetRequest  {: #declarativenetrequest-conditional-perms }

Most use cases for declarativeNetRequest don't require any host permissions at all. However, some
do.

{% Aside 'caution' %}

> Host permissions are still required if the extension wants to *redirect* a request or modify headers
> on it. The `declarativeNetRequestWithHostAccess` permission always requires host permissions to the
> request URL and initiator to act on a request.

Reviewer: Minor rephrasing for a grammar tweak. 

```suggestion
Host permissions are still required if the extension wants to *redirect* a request or modify its
headers. The `declarativeNetRequestWithHostAccess` permission always requires host permissions to
the request URL and initiator to act on a request.
```

{% endAside %}

When extensions require host permissions for these use cases, we recommend a "tiered" permissions
strategy. This means implementing the extension's core functionality without using these

> permissions; putting the more advanced use cases behind an `"optional_host_permission"`.

Reviewer: ```suggestion
permissions; putting the more advanced use cases behind an `"optional_host_permissions"` pattern.
```

This approach allows privacy-conscious users to withhold those permissions and still use much of the
extension's functionality. This means that developers can implement many common use cases, such as
content-blocking functionality, without requiring any host permissions.

## Sunset for deprecated APIs  {: #sunset-deprecated-apis }

There are a number of APIs that have long been deprecated. Manifest V3 finally removes support for

> the following deprecated APIs:

Reviewer: ```suggestion
the following deprecated methods and properties:
```

  func: showAlert,
  args: [name],
});

```
{% endColumns %}

A functional version of the Manifest V3 snippets in this section can be found in the
[chrome-extensions-samples][github-samples-content] repository. See the [Tabs API
examples][api-tabs-example] for an implementation of `getCurrentTab()`.

## Service workers  {: #background-service-workers }

Background pages in Manifest V2 are replaced by [service workers][dev-google-sw] in Manifest V3;
this is a foundational change that affects most extensions. The following are some notable
differences:

| MV2 - Background page       | MV3 - Service worker                            |
|-----------------------------|-------------------------------------------------|
| Can use a persistent page.  | Terminates when not in use.                     |
| Has access to the DOM.      | Doesn't have access to the DOM.                 |
| Can use `XMLHttpRequest()`. | Must use [fetch()][mdn-fetch] to make requests. |

See [Migrating from Background Pages to Service Workers][doc-background-to-worker] to explore how to
adapt to these and other challenges.

{% Aside %}

> In order to aid with the migration process, Manifest V2 extensions can use background service

Author: Good call! I'll make a separate PR for that 😀

or to trusted testers. 

### How do you use declarativeNetRequest?  {: #how-use-declarativenetrequest }

Instead of reading the request and programmatically altering it, your extension specifies a number
of [rules][api-declarative-rules]. Each rule contains a set of actions to perform when a given set
of conditions are matched. For example, you could define a rule that removes "cookie" headers when a
request is sent to a specific domain. See the [declarativeNetRequest][api-declarativenetrequest]
reference documentation for a more detailed description of rules

This feature allows content blockers and other request-modifying extensions to implement their use
cases without requiring host permissions, and without needing to read the actual requests.

### Conditional permissions and declarativeNetRequest  {: #declarativenetrequest-conditional-perms }

Most use cases for declarativeNetRequest don't require any host permissions at all. However, some
do.

{% Aside 'caution' %}

Host permissions are still required if the extension wants to *redirect* a request or modify its
headers. The `declarativeNetRequestWithHostAccess` permission always requires host permissions to

> the request URL and initiator to act on a request.

Reviewer 2: Should this be, "to he request URL and its initiator"? (Added the word 'its'.)

Author: Yep! Changing now
